### PR TITLE
llama : remove write/read of output ids/logits/embeddings

### PR DESCRIPTION
**Pull request by danbev**
This commit removes the write/read of output ids, logits and embeddings from the llama context state.

### CODE REVIEW

**Comment by ggerganov** [comment]
Initially, I was thinking that since the samplers can now be part of the context state, we should also store this information. But it should include also the sampler states. And it gets very complicated.

But now I am wondering if we should instead remove the output ids, the logits and the embeddings from the state and only store the model info and the memory. I can't think of a meaningful use case for storing this information. And even if its needed, one can simply run the last token through `llama_decode` to obtain the necessary logits/embeddings. So maybe this is the better option as it will simplify the read/write logic.

**Comment by danbev** [comment]
> And even if its needed, one can simply run the last token through `llama_decode` to obtain the necessary logits/embeddings.

I've updated [llama-completion](https://github.com/ggml-org/llama.cpp/pull/18862/commits/1d241c517858cab604e3a09938e1c6126ffd93e5) to "replay" the last token. Could you take a look and this if this is correctly handled for recurrent/hybrid models?

File: tools/completion/completion.cpp [hunk 1]
```diff
@@ -387,6 +387,32 @@ int main(int argc, char ** argv) {
...
+        // Logits are not stored as part of the session state so we need to
+        // "replay" the last token to get logits for sampling.
+        if (!session_tokens.empty() && n_match > 0 && n_match == session_tokens.size()) {
+            llama_token last_token = session_tokens.back();
+            int32_t pos;
+
+            if (llama_model_is_recurrent(model) || llama_model_is_hybrid(model)) {
+                LOG_INF("%s: recurrent/hybrid model: decode using next position: %d\n", __func__, (int)n_match);
+                pos = n_match; // use next position for decoding
+            } else {
```
Comment (ggerganov) [inline tools/completion/completion.cpp]: Hm this does not look right. It will evaluate token `n_match - 1` with position `n_match`.

Comment (danbev) [reply tools/completion/completion.cpp]: For recurrent models their hidden state is cumulative so we can't undo the last tokens effect on the state.
Would it perhaps be alright to store the logits just for recurrent/hybrid models? 
Or do you have any other alternatives for this that I might not have thought of?

Comment (ggerganov) [reply tools/completion/completion.cpp]: The solution here is to save the session at least one token before the full prompt is processed. We already do this in server:

https://github.com/ggml-org/llama.cpp/blob/eef375ce167cf36d8ed2cf347be95a7a7a90fb5e/tools/server/server-context.cpp#L2522-L2526

I looked at some point how to make a similar logic in `completion.cpp`, but seemed a bit complicated. Likely need to do some refactoring to simplify the logic first.

Ideally, we have to find a common pattern that is easy to apply - for example the `save-load-state.cpp` would need similar logic.

> Would it perhaps be alright to store the logits just for recurrent/hybrid models?

The few branches in the logic is always better, so prefer to avoid that.

File: src/llama-context.cpp [hunk 2]
```diff
@@ -2500,64 +2500,6 @@ size_t llama_context::state_write_data(llama_io_write_i & io) {
...
-        const auto & output_ids = this->output_ids;
-
-        std::vector<int32_t> w_output_pos;
-
-        w_output_pos.resize(n_outputs);
-
-        // build a more compact representation of the output ids
-        for (size_t i = 0; i < n_batch(); ++i) {
-            // map an output id to a position in the batch
-            int64_t pos = output_ids[i];
```
Comment (compilade) [inline src/llama-context.cpp:2499]: The main reason for output ids in the first place was when using `llama-perplexity` with e.g. the hellaswag benchmark, which uses parallel sequences of different lengths, which are re-ordered by the batch splitting logic for recurrent models to get equal-length sequences within ubatches.

(The initial reason was to make the output buffer smaller (see <https://github.com/ggml-org/llama.cpp/pull/6122>), but it allowed tackling batch splitting afterwards (https://github.com/ggml-org/llama.cpp/pull/8526))

I'm not sure whether simply replaying the last token will be sufficient in the case where more logits than that are required.

Comment (compilade) [reply src/llama-context.cpp:2499]: Whoops nevermind, this concern doesn't apply here, session saving doesn't affect things like `llama-perplexity`, I think.

Comment (danbev) [reply src/llama-context.cpp:2499]: I'll take a look just to make sure as I not familiar how llama-perplexity stores its information but I was under the impression that this format is different and not related.

File: common/common.cpp [hunk 3]
```diff
@@ -1863,3 +1863,66 @@ float lr_opt::get_lr(float epoch) const {
...
+}
+
+bool common_prompt_batch_decode(
+              struct llama_context * ctx,
+    const std::vector<llama_token> & tokens,
+                               int & n_past,
+                               int   n_batch,
+       const std::filesystem::path & state_path,
+                              bool   save_state,
+                              bool   is_last_batch) {
```
Comment (ggerganov) [inline common/common.cpp]: The API can be simpler:

```suggestion
bool common_prompt_batch_decode(
              struct llama_context * ctx,
    const std::vector<llama_token> & tokens,
                               int & n_past,
                               int   n_batch,
                 const std::string & state_fname) {
    const bool save_state = !state.fname.empty();
    ...
```

Don't think it's worth bringing `std::filesystem` here, so simple `std::string` should be fine.

File: common/common.h [hunk 4]
```diff
@@ -779,6 +779,23 @@ void common_batch_add(
...
+//
+// Note: We save state before the last token so that we can replay it to ensure
+// compatibility with all memory types. Recurrent/hybrid models cannot remove
+// tokens from memory, so this approach works across all model architectures.
+bool common_prompt_batch_decode(
+              struct llama_context * ctx,
+    const std::vector<llama_token> & embd,
+                               int & n_past,
+                               int   n_batch,
+            const std::string_view & state_path,
```
Comment (ggerganov) [inline common/common.h]: I think the preferred way in C++ usually is to pass `std::string_view` by value:


```suggestion
                  std::string_view   state_path,
```

**Review by ggerganov — APPROVED** [review]
I think the change is OK. Just double check that `save-load-state` works correctly with few different model types.

We should think how to exercise this in the CI. If we are able to generate small test models of different architectures as Johannes mentioned in https://github.com/ggml-org/llama.cpp/pull/19378#issuecomment-3899326129, we can add extra `save-load-state` and `server` tests to the CI in order to cover recurrent and hybrid arches.

**Comment by danbev** [comment]
@ggerganov I've tested `save-load-state` with recurrent, hybrid, and standard transformer models.  

I'll take a look at https://github.com/ggml-org/llama.cpp/pull/19802 and follow up regarding adding tests for this to CI.

**Comment by fairydreaming** [comment]
@danbev if you have some time please take a look at #23400
